wolfcrypt/src/sakke.c

```diff
@@ -0,0 +1,6774 @@
+ *
+ * @param  [in]   key     SAKKE key.
+ * @param  [out]  authSz  Size of authentication data in bytes.
+ * @return  0 on success.
+ * @return  BAD_FUNC_ARG when key or authSz is NULL.
+ * @return  MEMORY_E when dynamic memory allocation fails.
+ * @return  Other -ve value on internal failure.
+ */
+int wc_GetSakkeAuthSize(SakkeKey* key, word16* authSz)
+{
+    int err = 0;
+
+    if ((key == NULL) || (authSz == NULL)) {
+        err = BAD_FUNC_ARG;
+    }
+
+    if (err == 0) {
+        err = sakke_load_params(key);
+    }
+    if (err == 0) {
+        word16 n = (word16)((mp_count_bits(&key->params.prime) + 7) / 8);
+        *authSz = 1 + 2 * n;
+    }
+
+    return err;
+}
+
+#ifdef WOLFSSL_HAVE_SP_ECC
+/*
+ * Modular exponentiate the value in F_p*.
+ *
+ * @param  [in]   key  SAKKE key.
+ * @param  [in]   b    MP integer that is the base to exponentiate.
+ * @param  [in]   e    MP integer that is the exponent.
+ * @param  [out]  r    Result of exponentiation.
+ * @return  0 on success.
+ * @return  MEMORY_E when dynamic memory allocation fails.
+ * @return  Other -ve value on internal failure.
+ */
+static int sakke_modexp(const SakkeKey* key, const mp_int* b, mp_int* e,
+        mp_int* r)
+{
+    int err = NOT_COMPILED_IN;
+
+#ifdef WOLFSSL_SP_1024
+    if ((key->ecc.idx != ECC_CUSTOM_IDX) &&
+            (ecc_sets[key->ecc.idx].id == ECC_SAKKE_1)) {
+        err = sp_ModExp_Fp_star_1024(b, e, r);
+    }
+#endif
+
+    return err;
+}
+#else
+#ifdef WOLFSSL_SMALL
+/*
+ * Modular exponentiate the value in F_p*.
+ *
+ * @param  [in]   key  SAKKE key.
+ * @param  [in]   b    MP integer that is the base to exponentiate.
+ * @param  [in]   e    MP integer that is the exponent.
+ * @param  [out]  c    Result of exponentiation.
+ * @param  [in]   mp   Multiplier to use when converting from Montgomery form.
+ * @return  0 on success.
+ * @return  MEMORY_E when dynamic memory allocation fails.
+ * @return  Other -ve value on internal failure.
+ */
+static int sakke_modexp_loop(SakkeKey* key, mp_int* b, mp_int* e, mp_proj* c,
+        mp_digit mp)
+{
+    int err;
+    mp_int* t1 = &key->tmp.m1;
+    mp_int* t2 = &key->tmp.m2;
+    mp_int* by = key->tmp.p1->z;
+    mp_int* prime = &key->params.prime;
+    int i;
+
+    /* Set the working value to the base in PF_p[q] */
+    err = mp_montgomery_calc_normalization(c->x, prime);
+    /* Set c->y is montgomery form of b - base */
+    if (err == 0) {
+        err = mp_mulmod(b, c->x, prime, by);
+    }
+    /* Set by is montgomery form of b - base */
+    if (err == 0) {
+        err = mp_copy(by, c->y);
+    }
+    for (i = mp_count_bits(e) - 2; (err == 0) && (i >= 0); i--) {
+        err = sakke_proj_sqr(c, prime, mp, c, t1, t2);
+        if (err == 0) {
+            if (mp_is_bit_set(e, i)) {
+                err = sakke_proj_mul_qx1(c, by, prime, mp, c, t1, t2);
+            }
+        }
+    }
+
+    return err;
+}
+
+#elif defined(WOLFSSL_SAKKE_SMALL_MODEXP) || !defined(USE_FAST_MATH)
```

Comment:
Can you add a comment at top for `WOLFSSL_SAKKE_SMALL_MODEXP`? Also made note to add Jenkins test case for this.

Reply:
Done.